horovod/mxnet/mpi_ops.cc

```diff
@@ -49,206 +51,154 @@ inline void InvokeCompleteCallback(CallbackOnComplete on_complete, const Status&
-      };
+      enqueue_result = EnqueueTensorAllgather(
+          hvd_context, hvd_tensor, nullptr, name, device,
+          [on_complete](const Status& status) {
+            InvokeCompleteCallback(on_complete, status);
+      });
       break;
     case OperationType::BROADCAST:
-      op_type_name = "horovod_broadcast";
-      op_name = GetOpName(op_type_name, name);
-      exec_fn = [input, output, op_name, root_rank]
-                (RunContext rctx, CallbackOnComplete on_complete) mutable {
-        DoBroadcast(input, output, op_name, root_rank, on_complete);
-      };
+      if (horovod_rank() == ops_param->root_rank) {
+        if (tensor != output) {
+          TensorUtil::Copy(output, tensor);
+        }
+      } else {
+        hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+      }
+
+      enqueue_result = EnqueueTensorBroadcast(
+          hvd_context, hvd_tensor, hvd_output, ops_param->root_rank,
+          nullptr, name, device,
+          [on_complete](const Status& status) {
+            InvokeCompleteCallback(on_complete, status);
+      });
       break;
     default:
-      LOG(FATAL) << "Unsupported Horovod operation type";
+      throw std::logic_error("Unsupported Horovod operation type.");
   }
 
+  ThrowIfError(enqueue_result);
+}
+
+inline void PushHorovodOperation(OperationType op_type, NDArray* input,
+                                 NDArray* output, const char* name,
+                                 int priority, int root_rank = -1) {
+  auto op_type_name = GetOpTypeName(op_type);
+  auto op_name = GetOpName(op_type_name, name);
+  auto ops_param = CreateMpiOpsParam(input, output, op_type, op_name, root_rank, false);
+
   // Not in-place
-  if (input->var() != output->var()) {
-    Engine::Get()->PushAsync(exec_fn, MX_EXEC_CTX,
-                             {input->var()}, {output->var()},
-                             FnProperty::kCPUPrioritized, priority,
-                             op_type_name.c_str());
+  auto input_var = input->var();
+  auto output_var = output->var();
+  if (input_var != output_var) {
+    MXEnginePushAsync(DoHorovodOperation, ops_param, DeleteMpiOpsParam,
+                      &MX_EXEC_CTX, &input_var, 1, &output_var, 1,
+                      &MX_FUNC_PROP, priority, op_type_name);
   // In-place
   } else {
-    Engine::Get()->PushAsync(exec_fn, MX_EXEC_CTX,
-                             {}, {output->var()},
-                             FnProperty::kCPUPrioritized, priority,
-                             op_type_name.c_str());
+    MXEnginePushAsync(DoHorovodOperation, ops_param, DeleteMpiOpsParam,
+                      &MX_EXEC_CTX, nullptr, 0, &output_var, 1,
+                      &MX_FUNC_PROP, priority, op_type_name);
   }
 }
 
 #if HAVE_CUDA
-void DoAllreduceCudaOnCPU(MXTempBufferShared& hvd_cpu_buffer, std::string& name,
-                          CallbackOnComplete on_complete) {
+void DoHorovodOperationCudaOnCPU(void*, void* on_complete_ptr, void* param) {
```

Comment:
Do we still need the first argument in the function?

Reply:
We need it to comply with the [function signature](https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/c_api.h#L106). The first argument is RunContext, which we do not use here but might be used in other cases. That's why the function signature has it to make it more generic.